relayer/chains/substrate/tx.go

```diff
@@ -0,0 +1,901 @@
+				if err != nil {
+					return nil, nil, err
+				}
+				return nil, timeout, nil
+			} else {
+				timeout, err := src.MsgTimeout(msgTransfer, pp)
+				if err != nil {
+					return nil, nil, err
+				}
+				return nil, timeout, nil
+			}
+		default:
+			return nil, nil, err
+		}
+	}
+
+	pp, err := src.PacketCommitment(ctx, msgTransfer, srch)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	packet, err := sp.MsgRecvPacket(msgTransfer, pp)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return packet, nil, nil
+}
+
+// AcknowledgementFromSequence relays an acknowledgement with a given seq on src, source is the sending chain, destination is the receiving chain
+func (sp *SubstrateProvider) AcknowledgementFromSequence(ctx context.Context, dst provider.ChainProvider, dsth, seq uint64, dstChanId, dstPortId, srcChanId, srcPortId string) (provider.RelayerMessage, error) {
+	msgRecvPacket, err := dst.QueryRecvPacket(ctx, dstChanId, dstPortId, seq)
+	if err != nil {
+		return nil, err
+	}
+
+	pp, err := dst.PacketAcknowledgement(ctx, msgRecvPacket, dsth)
+	if err != nil {
+		return nil, err
+	}
+	msg, err := sp.MsgAcknowledgement(msgRecvPacket, pp)
+	if err != nil {
+		return nil, err
+	}
+	return msg, nil
+}
+
+// QueryIBCHeader returns the IBC compatible block header (SubstrateIBCHeader) at a specific height.
+func (sp *SubstrateProvider) QueryIBCHeader(ctx context.Context, h int64) (provider.IBCHeader, error) {
+	if h <= 0 {
+		return nil, fmt.Errorf("must pass in valid height, %d not valid", h)
+	}
+
+	latestBeefyBlockHash, err := sp.RPCClient.RPC.Beefy.GetFinalizedHead()
+	if err != nil {
+		return nil, err
+	}
+
+	latestBeefyHeight, err := sp.RPCClient.RPC.Chain.GetBlock(latestBeefyBlockHash)
+	if err != nil {
+		return nil, err
+	}
+
+	if h > int64(latestBeefyHeight.Block.Header.Number) {
+		return nil, fmt.Errorf("queried block is not finalized")
+	}
+
+	blockHash, err := sp.RPCClient.RPC.Chain.GetBlockHash(uint64(h))
+	if err != nil {
+		return nil, err
+	}
+
+	header, err := sp.constructBeefyHeader(latestBeefyBlockHash, &blockHash)
+	if err != nil {
+		return nil, err
+	}
+
+	return SubstrateIBCHeader{
+		height:       uint64(h),
+		SignedHeader: header,
+	}, nil
+}
+
+// InjectTrustedFields injects the necessary trusted fields for a header to update a light
+// client stored on the destination chain, using the information provided by the source
+// chain.
+// TrustedHeight is the latest height of the IBC client on dst
+// TrustedValidators is the validator set of srcChain at the TrustedHeight
+// InjectTrustedFields returns a copy of the header with TrustedFields modified
+func (sp *SubstrateProvider) InjectTrustedFields(ctx context.Context, header ibcexported.Header, dst provider.ChainProvider, dstClientId string) (ibcexported.Header, error) {
+	// make copy of header stored in mop
+	h, ok := header.(*beefyclienttypes.Header)
+	if !ok {
+		return nil, fmt.Errorf("trying to inject fields into non-tendermint headers")
```

Comment:
This error message should be adjust to something that makes sense for Polkadot/Substrate